src/test/java/com/atlassian/db/replica/api/KeyBasedThrottledCacheTest.java

```diff
@@ -0,0 +1,110 @@
+package com.atlassian.db.replica.api;
+
+import org.junit.Test;
+
+import java.util.ConcurrentModificationException;
+import java.util.Optional;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class KeyBasedThrottledCacheTest {
+    private final ExecutorService executor = Executors.newFixedThreadPool(10);
+
+    @Test
+    public void staleWhileInvalidating() throws InterruptedException {
+        KeyBasedThrottledCache cache = new KeyBasedThrottledCache(randomKey());
+        Long staleValue = 1L;
+        cache.get(() -> staleValue);
+
+        asyncCacheUpdateWithSlowSupplier(cache, 2L);
+
+        Optional<Long> actual = cache.get(() -> 3L);
+
+        assertThat(actual).hasValue(staleValue);
+    }
+
+
+    @Test
+    public void serveLatestValue() {
+        KeyBasedThrottledCache cache = new KeyBasedThrottledCache(randomKey());
+        cache.get(() -> 1L);
+        cache.get(() -> 3L);
+
+        Long lastValue = 4L;
+        Optional<Long> actual = cache.get(() -> lastValue);
+
+        assertThat(actual).hasValue(lastValue);
+    }
+
+    @Test
+    public void noConcurrentInvalidation() throws InterruptedException {
```

Comment:
I don't caught the idea of this unit test. What is the assertion here? That there is no exception and empty optional is returned?

Reply:
Exactly. We can make it explicit via AssertJ: `Assertions.catchThrowable`

Reply:
It's pretty much proving that no concurrent invalidation is happening. If you scrap the locking it will fail.